File: src/arch/x86_64/gdt.rs
// x86_64 (64-bit) GDT implementation
// 64-bit long mode requires a different GDT layout

use core::arch::asm;

#[repr(C, packed)]
#[derive(Copy, Clone)]
struct GdtEntry {
    limit_low: u16,
    base_low: u16,
    base_middle: u8,
    access: u8,
    granularity: u8,
    base_high: u8,
}

#[repr(C, packed(2))]
struct GdtPtr {
    limit: u16,
    base: u64,
}

/// AMD64 TSS structure
#[repr(C, packed)]
pub struct Tss {
    reserved1: u32,
    pub rsp0: u64,
    pub rsp1: u64,
    pub rsp2: u64,
    reserved2: u64,
    ist1: u64,
    ist2: u64,
    ist3: u64,
    ist4: u64,
    ist5: u64,
    ist6: u64,
    ist7: u64,
    reserved3: u64,
    reserved4: u16,
    iomap_base: u16,
}

impl Tss {
    pub const fn new() -> Self {
        Tss {
            reserved1: 0,
            rsp0: 0,
            rsp1: 0,
            rsp2: 0,
            reserved2: 0,
            ist1: 0, ist2: 0, ist3: 0, ist4: 0,
            ist5: 0, ist6: 0, ist7: 0,
            reserved3: 0,
            reserved4: 0,
            iomap_base: 104,
        }
    }
}

static mut TSS: Tss = Tss::new();

// GDT with typical entries for long mode:
// Index 0: Null
// Index 1: Kernel Code (64-bit)
// Index 2: Kernel Data
// Index 3: User Code (64-bit)
// Index 4: User Data
// Index 5: TSS descriptor (16 bytes, spanning 2 entries)
static mut GDT: [GdtEntry; 7] = [GdtEntry {
    limit_low: 0, base_low: 0, base_middle: 0, access: 0, granularity: 0, base_high: 0,
}; 7];

static mut GDT_PTR: GdtPtr = GdtPtr { limit: 0, base: 0 };

pub unsafe fn init() {
    // 1. Null descriptor
    GDT[0] = GdtEntry { limit_low: 0, base_low: 0, base_middle: 0, access: 0, granularity: 0, base_high: 0 };

    // 2. Kernel Code (0x08) - 64-bit code segment
    // Access: Present, Ring 0, Code, Execute/Read
    // Granularity: Long mode, 64-bit
    GDT[1] = GdtEntry {
        limit_low: 0, base_low: 0, base_middle: 0,
        access: 0x9A, // P=1, DPL=0, S=1, Type=Code (Execute/Read)
        granularity: 0x20, // L=1 (Long mode), D/B=0
        base_high: 0,
    };

    // 3. Kernel Data (0x10)
    GDT[2] = GdtEntry {
        limit_low: 0xFFFF, base_low: 0, base_middle: 0,
        access: 0x92, // P=1, DPL=0, S=1, Type=Data (Read/Write)
        granularity: 0xCF,
        base_high: 0,
    };

    // 4. User Code (0x1B) - 64-bit code segment
    GDT[3] = GdtEntry {
        limit_low: 0, base_low: 0, base_middle: 0,
        access: 0xFA, // P=1, DPL=3, S=1, Type=Code (Execute/Read)
        granularity: 0x20, // L=1
        base_high: 0,
    };

    // 5. User Data (0x23)
    GDT[4] = GdtEntry {
        limit_low: 0xFFFF, base_low: 0, base_middle: 0,
        access: 0xF2, // P=1, DPL=3, S=1, Type=Data (Read/Write)
        granularity: 0xCF,
        base_high: 0,
    };

    // 6-7. TSS Descriptor (0x28, uses 2 entries = 16 bytes)
    let tss_base = &raw const TSS as u64;
    let tss_limit = (core::mem::size_of::<Tss>() - 1) as u32;

    // First 8 bytes of TSS descriptor
    GDT[5] = GdtEntry {
        limit_low: (tss_limit & 0xFFFF) as u16,
        base_low: (tss_base & 0xFFFF) as u16,
        base_middle: ((tss_base >> 16) & 0xFF) as u8,
        access: 0x89, // Present, Ring 0, TSS (Available)
        granularity: ((tss_limit >> 16) & 0x0F) as u8 | 0x00, // 1-byte granular
        base_high: ((tss_base >> 24) & 0xFF) as u8,
    };

    // Second 8 bytes of TSS descriptor (bits 32-63 of base and upper bits)
    // For GDT[6], we encode the upper 32 bits of the base
    // In standard format, this is stored as a separate 8-byte entry
    GDT[6] = GdtEntry {
        limit_low: ((tss_base >> 32) & 0xFFFF) as u16,
        base_low: ((tss_base >> 48) & 0xFFFF) as u16,
        base_middle: 0,
        access: 0,
        granularity: 0,
        base_high: 0,
    };

    GDT_PTR.limit = (core::mem::size_of::<[GdtEntry; 7]>() - 1) as u16;
    GDT_PTR.base = &raw const GDT as u64;

    let stack_ptr: u64;
    asm!("mov {}, rsp", out(reg) stack_ptr);
    TSS.rsp0 = stack_ptr;

    // Load GDT (64-bit version using lgdt with memory operand)
    asm!("lgdt [{}]", in(reg) &raw const GDT_PTR);

    // Reload segment registers
    asm!(
        "push 0x08",
        "lea rax, [2f + rip]",
        "push rax",
        "retfq",
        "2:",
        "mov ax, 0x10",
        "mov ds, ax",
        "mov es, ax",
        "mov fs, ax",
        "mov gs, ax",
        "mov ss, ax",
        out("rax") _,
        options(nostack)
    );

    // Load TSS via LTR
    asm!("ltr ax", in("ax") 0x28u16);
}

pub fn set_kernel_stack(addr: u64) {
    unsafe { TSS.rsp0 = addr; }
}

pub fn get_kernel_stack() -> u64 {
    unsafe { TSS.rsp0 }
}

File: kernel/x86_64/src/device/gdt.c
// NebulaOS - x86_64 GDT Implementation
// ======================================
//
// GDT implementation for x86_64 architecture

#include "../../../common/include/gdt.h"
#include "../../../common/include/nebula.h"

// GDT table (6 entries: null, kernel code, kernel data, user code, user data, TSS)
static gdt_entry_t gdt[6];

// GDT pointer
static gdt_ptr_t gdt_ptr;

// TSS entry
static tss_t tss;

// -----------------------------------------------------------------------------
// Initialize GDT for 64-bit long mode
// -----------------------------------------------------------------------------
void gdt64_init(void) {
    gdt_ptr.limit = sizeof(gdt) - 1;
    gdt_ptr.base = (uint64_t)&gdt;
    gdt_ptr.base_upper = (uint32_t)((uint64_t)&gdt >> 32);

    // NULL descriptor
    gdt_set_gate(0, 0, 0, 0, 0);

    // Kernel code segment (0x08) - 64-bit long mode
    gdt_set_gate(1, 0, 0,
                 GDT_ACCESS_PRESENT | GDT_ACCESS_PRIV_0 | GDT_ACCESS_TYPE_CODE | GDT_ACCESS_READABLE,
                 GDT_FLAGS_GRANULARITY | GDT_FLAGS_64BIT);

    // Kernel data segment (0x10) - flat data
    gdt_set_gate(2, 0, 0,
                 GDT_ACCESS_PRESENT | GDT_ACCESS_PRIV_0 | GDT_ACCESS_TYPE_DATA | GDT_ACCESS_WRITABLE,
                 GDT_FLAGS_GRANULARITY);

    // User code segment (0x18) - ring 3 code
    gdt_set_gate(3, 0, 0,
                 GDT_ACCESS_PRESENT | GDT_ACCESS_PRIV_3 | GDT_ACCESS_TYPE_CODE | GDT_ACCESS_READABLE,
                 GDT_FLAGS_GRANULARITY | GDT_FLAGS_64BIT);

    // User data segment (0x20) - ring 3 data
    gdt_set_gate(4, 0, 0,
                 GDT_ACCESS_PRESENT | GDT_ACCESS_PRIV_3 | GDT_ACCESS_TYPE_DATA | GDT_ACCESS_WRITABLE,
                 GDT_FLAGS_GRANULARITY);

    // TSS segment (0x28) - Task State Segment
    memset(&tss, 0, sizeof(tss_t));
    tss.ss0 = 0x10;
    tss.esp0 = 0;
    tss.iomap_base = sizeof(tss_t);

    // Set up TSS descriptor using a 64-bit compatible approach
    // For x86_64, the TSS descriptor is 16 bytes
    uint64_t tss_base = (uint64_t)&tss;
    uint64_t tss_limit = sizeof(tss_t) - 1;

    gdt[5].base_low = (uint16_t)(tss_base & 0xFFFF);
    gdt[5].base_mid = (uint8_t)((tss_base >> 16) & 0xFF);
    gdt[5].base_high = (uint8_t)((tss_base >> 24) & 0xFF);
    gdt[5].limit_low = (uint16_t)(tss_limit & 0xFFFF);
    gdt[5].access = GDT_ACCESS_PRESENT | GDT_ACCESS_PRIV_0 | 0x09;  // Available 64-bit TSS
    gdt[5].limit_high = (uint8_t)((tss_limit >> 16) & 0x0F);

    // Flush GDT
    gdt_flush64();
}

// -----------------------------------------------------------------------------
// Set GDT gate (shared with x86, handles flat 64-bit segments)
// -----------------------------------------------------------------------------
void gdt_set_gate(int32_t num, uint32_t base, uint32_t limit, uint8_t access, uint8_t flags) {
    gdt[num].base_low = (uint16_t)(base & 0xFFFF);
    gdt[num].base_mid = (uint8_t)((base >> 16) & 0xFF);
    gdt[num].base_high = (uint8_t)((base >> 24) & 0xFF);
    gdt[num].limit_low = (uint16_t)(limit & 0xFFFF);
    gdt[num].limit_high = (uint8_t)((flags & 0xF0) | ((limit >> 16) & 0x0F));
    gdt[num].access = access;
}

// -----------------------------------------------------------------------------
// Flush GDT for 64-bit
// Reload GDTR and all segment registers
// -----------------------------------------------------------------------------
void gdt_flush64(void) {
    __asm__ __volatile__(
        "lgdt %0\n"
        "mov $0x10, %%rax\n"
        "mov %%rax, %%ds\n"
        "mov %%rax, %%es\n"
        "mov %%rax, %%fs\n"
        "mov %%rax, %%gs\n"
        "mov %%rax, %%ss\n"
        "pushq $0x08\n"
        "lea .lgdt_return(%%rip), %%rax\n"
        "pushq %%rax\n"
        "lretq\n"
        ".lgdt_return:\n"
        : : "m" (gdt_ptr) : "rax", "memory"
    );
}

// -----------------------------------------------------------------------------
// Flush TSS (load TR register)
// -----------------------------------------------------------------------------
void tss_flush(void) {
    __asm__ __volatile__("ltr %%ax" : : "a" (GDT_TSS_SEL));
}
